### PR TITLE
servershell: Improve highlights for changes

### PR DESCRIPTION
**Pull request by YannikSc**
- Improve highlighting for changed items
- Add pin and unpin commands
- Changed and deleted items will be visible even if not matched by the search

### CODE REVIEW

File: serveradmin/servershell/static/js/servershell/command.js [hunk 1]
```diff
@@ -173,6 +173,23 @@ function validate_selected(min=1, max=-1) {
 }
 
 servershell.commands = {
+    pin: function() {
+        if (!Array.isArray(servershell.pinned)) {
+            servershell.pinned = [];
```
Comment (kofrezo) [inline serveradmin/servershell/static/js/servershell/command.js:178]: Lets move the definition and initialisation [here](https://github.com/innogames/serveradmin/blob/main/serveradmin/servershell/templates/servershell/index.html#L214) to be consistent.

File: serveradmin/servershell/static/js/servershell/command.js [hunk 2]
```diff
@@ -173,6 +173,23 @@ function validate_selected(min=1, max=-1) {
...
+        }
+
+        let selected_ids = servershell.get_selected();
+        servershell.pinned.push(...selected_ids);
+        servershell.pinned = servershell.pinned.filter((value, index) =>
```
Comment (kofrezo) [inline serveradmin/servershell/static/js/servershell/command.js:183]:
```suggestion
        servershell.pinned = [...new Set(servershell.pinned)];
```

File: serveradmin/servershell/static/js/servershell/search.js [hunk 3]
```diff
@@ -1,3 +1,60 @@
+/**
+ * Sends the ajax request to the serveradmin api
+ *
```
Comment (kofrezo) [inline serveradmin/servershell/static/js/servershell/search.js:3]:
```suggestion
 *
 * Send the search term to the backend, process and display the results or show errors if there are any.
```

File: serveradmin/servershell/views.py [hunk 4]
```diff
@@ -128,11 +128,53 @@ def autocomplete(request):
                         content_type='application/x-json')
 
 
+class MergedQuery:
```
Comment (kofrezo) [inline serveradmin/servershell/views.py:131]: This and the iterator class should go to a dedicated module/file.

Comment (kofrezo) [inline serveradmin/servershell/views.py:131]: A doc string for both would be nice please.

File: serveradmin/servershell/views.py [hunk 5]
```diff
@@ -156,12 +198,17 @@ def get_results(request):
         restrict = shown_attributes.copy()
         if 'servertype' not in restrict:
             restrict.append('servertype')
-        query = Query(parse_query(term), restrict, order_by)
+        main_query = Query(parse_query(term), restrict, order_by)
+
+        query = MergedQuery([
+            Query({'object_id': Any(*pinned)}, restrict, None),
```
Comment (kofrezo) [inline serveradmin/servershell/views.py:204]: You can ommit the third parameter if not used.
```suggestion
            Query({'object_id': Any(*pinned)}, restrict),
```

File: serveradmin/servershell/views.py [hunk 6]
```diff
@@ -156,12 +198,17 @@ def get_results(request):
         restrict = shown_attributes.copy()
         if 'servertype' not in restrict:
             restrict.append('servertype')
-        query = Query(parse_query(term), restrict, order_by)
+        main_query = Query(parse_query(term), restrict, order_by)
+
+        query = MergedQuery([
```
Comment (kofrezo) [inline serveradmin/servershell/views.py:203]: Probably better to call this `queries` or `merged_queries` or similar.

File: serveradmin/servershell/views.py [hunk 7]
```diff
@@ -156,12 +198,17 @@ def get_results(request):
...
+            Query({'object_id': Any(*pinned)}, restrict, None),
+            main_query,
+        ])
 
         # TODO: Using len is terribly slow for large datasets because it has
         #  to query all objects but we cannot use count which is available on
         #  Django QuerySet
-        num_servers = len(query)
+        num_servers = len(list(query))
```
Comment (kofrezo) [inline serveradmin/servershell/views.py:211]: Wouldn't this give us a list of Queries ?

**Review by kofrezo — CHANGES_REQUESTED** [review]
I added some comments.

Could you also please split the changes into multiple commits as well as make dedicated commits for reformatting code. It gets quite challenging to review all at once and not get lost.

File: serveradmin/servershell/views.py (hunk 7 above)
Comment (YannikSc) [reply serveradmin/servershell/views.py:211]: `list()` calls the `__iter__()` method on the `MergedQuery` so it gets a `MergedQueryIterator` and iterates over it to create a list out of its items.

**Review by kofrezo — APPROVED** [review]

**Pull request merged**